### ISSUE TITLE
Is this 2016 StackOverflow answer still accurate? (`fh = open(p); Lock(fh)` vs. `with Lock(p) as fh`)

### ISSUE THREAD

**Issue by joshuacwnewton** [wolph/portalocker#94]
Hi there! While trying to understand a separate issue, I came across [this SO answer](https://stackoverflow.com/questions/39292051/portalocker-does-not-seem-to-lock/39338942#39338942) that was peculiar to me.

----

Original question:

> > However, portalocker does not work in the simplest case. I created a simple file:
> > 
> > ```python
> > >>> f = open("test",'w')
> > >>> portalocker.lock(f, portalocker.LOCK_EX)
> > ```
> > 
> > Meanwhile I can still open it in another terminal:

Wolph's reponse:

> To make sure your code works properly in both cases I would suggest encapsulating both of the open calls with the locker.
> 
> ```python
> import portalocker
> 
> with portalocker.Lock('test') as fh:
>     fh.write('first instance')
>     print('waiting for your input')
>     input()
> ```

----

The distinction here seems to be the use of `Lock` to open the file (as a context manager), versus `open` then `lock`/`unlock`.

But, I notice this answer was written in 2016, so I wouldn't be surprised if `portalocker` has been updated since so that there is no longer such a distinction. And in fact, when we [took a look](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4423#issuecomment-2049903702) at the source...

> Looking at the code in [portalocker.utils::Lock](https://github.com/wolph/portalocker/blob/cdbefd4a5f272b8ad8c4d0468931e97d01854e62/portalocker/utils.py#L152-L167), I see that internally it does the same thing: open the file, then lock (and this is basically required by the OS interface, because you can only a lock on an open file handle). So, I don't think that's the problem.

Is the 2016 answer out of date, then? Or are we misunderstanding the distinction here?

Thank you kindly for maintaining this package! :heart:

**Comment by wolph** [comment]
The `with` statement adds a lot of convenience by opening, locking, closing and unlocking in the correct order. And it's a lot smarter when it comes to retrying, timeouts, etc...

The actual lock (when everything succeeds) is identical to `portalocker.lock(...)` so you can always use that as well. I always recommend the contextwrapper since it's just more convenient, 

The most important part of the answer to that SO question wasn't the contextlib vs direct lock call, it's the "advisory locks" part. On Windows, by default, locking is by default mandatory and exclusive. On Linux and most other systems this is not the case, so unless both applications are using locking code, it won't lock.

Simply put, if you do this from the first Python console:

```python
>>> f = open("test",'w')
>>> portalocker.lock(f, portalocker.LOCK_EX)
```

Than nothing will stop you from doing this from a second Python console:

```
>>> f = open("test",'w')
```

Only when you execute this line, you will get an error:

```
>>> portalocker.lock(f, portalocker.LOCK_EX)
```

**Comment by joshuacwnewton** [comment]
Thank you for your kind and patient explanation. I very much appreciate the thorough response! ♥️
